metagraph/src/annotation/row_diff_builder.cpp

```diff
@@ -378,34 +397,28 @@ void build_pred_succ(const std::string &graph_fname,
         std::vector<bool> pred_boundary_buf;
 
         for (uint64_t i = start; i < std::min(start + BS, graph.num_nodes() + 1); ++i) {
-            BOSS::edge_index boss_idx = graph.kmer_to_boss_index(i);
-            if (!dummy[boss_idx]) {
+            bool find_succ = true;
+            if (succinct) { // Legacy code for DBGSuccinct
+                const BOSS &boss = succinct->get_boss();
+                BOSS::edge_index boss_idx = succinct->kmer_to_boss_index(i);
                 BOSS::edge_index next = boss.fwd(boss_idx);
                 assert(next);
-                if (!dummy[next]) {
-                    while (rd_succ.size() && !rd_succ[next]) {
-                        next--;
-                        assert(!boss.get_last(next));
-                    }
-                    succ_buf.push_back(to_row(graph.boss_to_kmer_index(next)));
-                    succ_boundary_buf.push_back(0);
-                }
-                // compute predecessors only for row-diff successors
-                if (rd_succ.size() ? rd_succ[boss_idx] : boss.get_last(boss_idx)) {
-                    BOSS::TAlphabet d = boss.get_node_last_value(boss_idx);
-                    BOSS::edge_index back_idx = boss.bwd(boss_idx);
-                    boss.call_incoming_to_target(back_idx, d,
-                        [&](BOSS::edge_index pred) {
-                            // dummy predecessors are ignored
-                            if (!dummy[pred]) {
-                                uint64_t node_index = graph.boss_to_kmer_index(pred);
-                                pred_buf.push_back(to_row(node_index));
-                                pred_boundary_buf.push_back(0);
-                            }
-                        }
-                    );
+                if ((*dummy)[next]) {
+                    find_succ = false;
                 }
             }
+            if(find_succ) {
+                auto j = graph.row_diff_successor(i, rd_succ);
+                succ_buf.push_back(to_row(j));
+                succ_boundary_buf.push_back(0);
+            }
+            if(rd_succ[from_graph_index(graph, i)]) {
+                graph.adjacent_incoming_nodes(i, [&](auto pred) {
+                    pred_buf.push_back(to_row(pred));
+                    pred_boundary_buf.push_back(0);
+                });
+            }
```

Comment:
I still think it should be fine to do it this way even we work with DBGSuccinct. What I mean is, as I understand it, dummy nodes are supposed to be something BOSS-internal, and not exposed in DeBruijnGraph interface, right? Then, if a dummy node is returned from `adjacent_incoming_nodes`, shouldn't it be considered a bug of DBGSuccinct, rather than smt that needs to be bypassed in this specific place?

On a side note, it looks like `pred` here is never a dummy node here, at least in `LabeledAlignerTest` group.